File: nakayama.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""6 nimmt! solver"""

import sys, random

# count cattle heads in cards
def count_cattleheads(cards):
	heads = 0
	for card in cards:
		if card == 55: heads += 7
		elif card % 11 == 0: heads += 5
		elif card % 10 == 0: heads += 3
		elif card % 5 == 0: heads += 2
		else: heads += 1
	return heads

# choose a card in hand
def solve(hand, rows):
	card = random.choice(hand)
	hand.remove(card)
	return card

# select a row of cards to take
def select(rows):
	return random.randint(0, len(rows) - 1)

def main(args):
	N = -1
	names = []
	hand = []
	while True:
		data = sys.stdin.readline().strip().split()
		status = data[0]
		if status == "players":
			N = int(data[1])
			names = data[2:]
			print
			sys.stdout.flush()
		elif status == "cards":
			num_cards = int(data[1])
			hand = sorted(map(int, data[2].split(",")))
			print
			sys.stdout.flush()
			break

	if N != len(names):
		print >>sys.stderr, "Error: incorrect number of players."
	if len(hand) != 10:
		print >>sys.stderr, "Error: incorrect number of cards in hand."

	while True:
		data = sys.stdin.readline().strip().split()
		status = data[0]
		if status == "rows":
			rows = [map(int, d.split(",")) for d in data[1:]]
			# send card from the hand via standard output
			print solve(hand, rows)
		elif status == "places":
			# placed cards of each players
			places = [map(int, d.split(" ")) for d in data[1:]]
			print
		elif status == "info":
			rows = [map(int, d.split(",")) for d in data[1:5]]
			opp = int(data[5])       # player index
			opp_card = int(data[6])  # placed card by player
			opp_idx = int(data[7])   # index of selected row by player
			if len(data) == 9:
				opp_row = map(int, data[8].split(","))  # taken cards by player
			print
		elif status == "select":
			rows = [map(int, d.split(",")) for d in data[1:]]
			# send index of selected row via standard output
			print select(rows)
		elif status == "quit":
			print
			break
		sys.stdout.flush()

if __name__ == "__main__": main(sys.argv)

File: nakayama.rb
#!/usr/bin/env ruby
# -*- mode:ruby; coding:utf-8 -*-

def count_cattleheads(cards)
  heads = 0
  cards.each{|card|
    if card == 55 then heads += 7
    elsif card % 11 == 0 then heads += 5
    elsif card % 10 == 0 then heads += 3
    elsif card % 5 == 0 then heads += 2
    else heads += 1
    end
  }
  return heads
end

def solve(hand, rows)
  card = hand.sample
  hand.delete(card)
  return card
end

def select(rows)
  heads_of_rows = rows.map{|row| count_cattleheads(row)}
  debug(heads_of_rows, "heads_of_rows")
  idx = heads_of_rows.index(heads_of_rows.min)
  debug(idx, "idx")
  return idx
end

def main(args)
  n = -1
  names = []
  hand = []
  while true do
    data = gets.strip().split()
    debug(data, "data")
    status = data[0]
    if status == "players"
      n = data[1].to_i
      data.shift(2)
      names = data
      debug(names,"names")
      print
      $stdout.flush
    elsif status == "cards"
      num_cards = data[1].to_i
      hand = data[2].split(",").map{|x| x.to_i}.sort
      print
      $stdout.flush
      debug(hand, "hand")
      break
    end
  end
  
  if n != names.size then
    $stderr.puts("Error: incorrect number of players.ruby")
  end
  if hand.size != 10 then
    $stderr.puts("Error: incorrect number of cards in hand.ruby")
  end
  
  while true do
    data = gets.strip().split()
    debug(data, "data")
    status = data[0]
    if status == "rows"
      rows = []
      data.shift(1)
      debug(data,"data")
      data.each{|row|
       rows.push(row.split(",").map{|x| x.to_i})
      }
      # debug(rows,"rows")
      puts solve(hand, rows)
    elsif status == "places"
      places = []
      data.shift(1)
      data.each{|row|
       places.push(row.split(",").map{|x| x.to_i})
      }
      puts
    elsif status == "info"
      rows = []
      data.shift(1)
      data[1..4].each{|row|
       rows.push(row.split(",").map{|x| x.to_i})
      }
      debug(rows, "rows")
      opp = data[5].to_i
      opp_card = data[6].to_i
      opp_idx = data[7].to_i
      if data.size == 9
        opp_row = data[8].split(",").map{|x| x.to_i}
        debug(opp_row, "opp_row")
      end
      puts
    elsif status == "select"
      rows = []
      data.shift(1)
      data.each{|row|
       rows.push(row.split(",").map{|x| x.to_i})
      }
      debug(rows, "rows")
      puts select(rows)
    elsif status == "quit"
      puts
      break
    end
    $stdout.flush
  end
end

def debug(arg, name)
  $stderr.puts "#{name} is.."
  $stderr.print arg
  $stderr.puts
end
    
if __FILE__ == $0 then
  main(ARGV)
end
